src/AspNet.Security.OAuth.Weixin/WeixinAuthenticationHandler.cs

```diff
@@ -109,13 +126,20 @@ protected override async Task<OAuthTokenResponse> ExchangeCodeAsync(string code,
 
         protected override string BuildChallengeUrl(AuthenticationProperties properties, string redirectUri)
         {
+            var oauthstate = Options.StateDataFormat.Protect(properties);
+            if (Options.AuthorizationEndpoint != WeixinAuthenticationDefaults.AuthorizationEndpoint)
+            {
+                //Store state in redirectUri when authorizing Wechat Web pages to prevent "too long state parameters" error
+                redirectUri = redirectUri.Contains("?") ? $"{redirectUri}&{nameof(oauthstate)}={oauthstate}" : $"{redirectUri}?{nameof(oauthstate)}={oauthstate}";
```

Comment:
Is this safe against URI encoding issues?

Reply:
Suddenly understand this problem, what good suggestions do you have?

Reply:
You could use [`Uri.EscapeDataString(string)`](https://docs.microsoft.com/en-us/dotnet/api/system.uri.escapedatastring?view=netcore-2.2) or [`Uri.EscapeUriString(string)`](https://docs.microsoft.com/en-us/dotnet/api/system.uri.escapeuristring?view=netcore-2.2) (I forget which is the right one).